In [10]:
import tensorflow as tf
from tensorflow.examples.tutorials.mnist import input_data

#load data
mnist = input_data.read_data_sets("MNIST_data", one_hot=True)

batch_size = 100

batch_count = mnist.train.num_examples // batch_size

#define place holder
x = tf.placeholder(tf.float32, [None, 784])
y = tf.placeholder(tf.float32, [None, 10])
keep_prob = tf.placeholder(tf.float32)

#create a 4-level neural network
W1 = tf.Variable(tf.truncated_normal([784, 2000], stddev=0.1))
b1 = tf.Variable(tf.zeros([2000])+0.1)
L1 = tf.nn.tanh(tf.matmul(x, W1)+b1)
L1_dropout = tf.nn.dropout(L1, keep_prob)

W2 = tf.Variable(tf.random_normal([2000, 2000], stddev=0.1))
b2 = tf.Variable(tf.zeros([2000])+0.1)
L2 = tf.nn.tanh(tf.matmul(L1_dropout, W2)+b2)
L2_dropout = tf.nn.dropout(L2, keep_prob)

W3 = tf.Variable(tf.random_normal([2000, 1000], stddev=0.1))
b3 = tf.Variable(tf.zeros([1000])+0.1)
L3 = tf.nn.tanh(tf.matmul(L2_dropout, W3)+b3)
L3_dropout = tf.nn.dropout(L3, keep_prob)

W4 = tf.Variable(tf.random_normal([1000, 10], stddev=0.1))
b4 = tf.Variable(tf.zeros([10])+0.1)
prediction = tf.nn.tanh(tf.matmul(L3_dropout, W4)+b4)


#二次代价函数
# loss = tf.reduce_mean(tf.square(y - prediction))
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y, logits=prediction))
# loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=y, logits=prediction))

#train
train = tf.train.GradientDescentOptimizer(0.2).minimize(loss)

init = tf.global_variables_initializer()

#比较预测值与实际值
correct_pred = tf.equal(tf.argmax(y, 1), tf.argmax(prediction, 1)) #argmax 返回沿着某个维度最大值的位置

#准确率
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

with tf.Session() as sess:
    sess.run(init)
    
    print("keep_prob 1.0")
    for step in range(20):
        for batch in range(batch_count):
            batch_xs, batch_ys = mnist.train.next_batch(batch_size)
            sess.run(train, feed_dict={x:batch_xs, y:batch_ys, keep_prob:1.0})
        
        test_acc = sess.run(accuracy, feed_dict={x:mnist.test.images, y:mnist.test.labels, keep_prob:1.0})
        train_acc = sess.run(accuracy, feed_dict={x:mnist.train.images, y:mnist.train.labels, keep_prob:1.0})
        print("step:" + str(step) + " test accurancy:" + str(test_acc) + " train accurancy:" + str(train_acc))

    print("keep_prob 0.7")
    for step in range(20):
        for batch in range(batch_count):
            batch_xs, batch_ys = mnist.train.next_batch(batch_size)
            sess.run(train, feed_dict={x:batch_xs, y:batch_ys, keep_prob:0.7})
        
        test_acc = sess.run(accuracy, feed_dict={x:mnist.test.images, y:mnist.test.labels, keep_prob:1.0})
        train_acc = sess.run(accuracy, feed_dict={x:mnist.train.images, y:mnist.train.labels, keep_prob:1.0})
        print("step:" + str(step) + " test accurancy:" + str(test_acc) + " train accurancy:" + str(train_acc))

    print("keep_prob 0.5")
    for step in range(20):
        for batch in range(batch_count):
            batch_xs, batch_ys = mnist.train.next_batch(batch_size)
            sess.run(train, feed_dict={x:batch_xs, y:batch_ys, keep_prob:0.5})
        
        test_acc = sess.run(accuracy, feed_dict={x:mnist.test.images, y:mnist.test.labels, keep_prob:1.0})
        train_acc = sess.run(accuracy, feed_dict={x:mnist.train.images, y:mnist.train.labels, keep_prob:1.0})
        print("step:" + str(step) + " test accurancy:" + str(test_acc) + " train accurancy:" + str(train_acc))


Extracting MNIST_data\train-images-idx3-ubyte.gz
Extracting MNIST_data\train-labels-idx1-ubyte.gz
Extracting MNIST_data\t10k-images-idx3-ubyte.gz
Extracting MNIST_data\t10k-labels-idx1-ubyte.gz
keep_prob 1.0
step:0 test accurancy:0.9487 train accurancy:0.95296365
step:1 test accurancy:0.9584 train accurancy:0.9706909
step:2 test accurancy:0.9617 train accurancy:0.9770182
step:3 test accurancy:0.9701 train accurancy:0.9845091
step:4 test accurancy:0.971 train accurancy:0.98765457
step:5 test accurancy:0.9713 train accurancy:0.98987275
step:6 test accurancy:0.9739 train accurancy:0.9916
step:7 test accurancy:0.9736 train accurancy:0.9924545
step:8 test accurancy:0.9748 train accurancy:0.9929818
step:9 test accurancy:0.9755 train accurancy:0.9931273
step:10 test accurancy:0.9751 train accurancy:0.99332726
step:11 test accurancy:0.9747 train accurancy:0.99347275
step:12 test accurancy:0.9758 train accurancy:0.99347275
step:13 test accurancy:0.9755 train accurancy:0.99349093
step:14 test ac

Extracting MNIST_data\train-images-idx3-ubyte.gz
Extracting MNIST_data\train-labels-idx1-ubyte.gz
Extracting MNIST_data\t10k-images-idx3-ubyte.gz
Extracting MNIST_data\t10k-labels-idx1-ubyte.gz
keep_prob 1.0
step:0 test accurancy:0.9487 train accurancy:0.95296365
step:1 test accurancy:0.9584 train accurancy:0.9706909
step:2 test accurancy:0.9617 train accurancy:0.9770182
step:3 test accurancy:0.9701 train accurancy:0.9845091
step:4 test accurancy:0.971 train accurancy:0.98765457
step:5 test accurancy:0.9713 train accurancy:0.98987275
step:6 test accurancy:0.9739 train accurancy:0.9916
step:7 test accurancy:0.9736 train accurancy:0.9924545
step:8 test accurancy:0.9748 train accurancy:0.9929818
step:9 test accurancy:0.9755 train accurancy:0.9931273
step:10 test accurancy:0.9751 train accurancy:0.99332726
step:11 test accurancy:0.9747 train accurancy:0.99347275
step:12 test accurancy:0.9758 train accurancy:0.99347275
step:13 test accurancy:0.9755 train accurancy:0.99349093
step:14 test accurancy:0.9756 train accurancy:0.99356365
step:15 test accurancy:0.9755 train accurancy:0.99356365
step:16 test accurancy:0.9754 train accurancy:0.9936
step:17 test accurancy:0.976 train accurancy:0.99365455
step:18 test accurancy:0.976 train accurancy:0.9936909
step:19 test accurancy:0.9762 train accurancy:0.9937091
keep_prob 0.7
step:0 test accurancy:0.9527 train accurancy:0.9570364
step:1 test accurancy:0.9517 train accurancy:0.95541817
step:2 test accurancy:0.954 train accurancy:0.9589091
step:3 test accurancy:0.9622 train accurancy:0.9636
step:4 test accurancy:0.9587 train accurancy:0.96369094
step:5 test accurancy:0.9611 train accurancy:0.9657818
step:6 test accurancy:0.9623 train accurancy:0.9663091
step:7 test accurancy:0.9647 train accurancy:0.9682182
step:8 test accurancy:0.9653 train accurancy:0.96976364
step:9 test accurancy:0.9652 train accurancy:0.9700364
step:10 test accurancy:0.9675 train accurancy:0.9715091
step:11 test accurancy:0.9664 train accurancy:0.9728
step:12 test accurancy:0.967 train accurancy:0.97285455
step:13 test accurancy:0.969 train accurancy:0.97423637
step:14 test accurancy:0.9687 train accurancy:0.9736182
step:15 test accurancy:0.97 train accurancy:0.9746909
step:16 test accurancy:0.9692 train accurancy:0.9759455
step:17 test accurancy:0.9712 train accurancy:0.97654545
step:18 test accurancy:0.971 train accurancy:0.9767454
step:19 test accurancy:0.9713 train accurancy:0.97803634
keep_prob 0.5
step:0 test accurancy:0.9621 train accurancy:0.96778184
step:1 test accurancy:0.9608 train accurancy:0.9642909
step:2 test accurancy:0.9615 train accurancy:0.9647273
step:3 test accurancy:0.9626 train accurancy:0.96541816
step:4 test accurancy:0.9611 train accurancy:0.9637455
step:5 test accurancy:0.9626 train accurancy:0.96427274
step:6 test accurancy:0.9621 train accurancy:0.9645636
step:7 test accurancy:0.963 train accurancy:0.96567273
step:8 test accurancy:0.9618 train accurancy:0.9643091
step:9 test accurancy:0.9615 train accurancy:0.9651091
step:10 test accurancy:0.9629 train accurancy:0.96565455
step:11 test accurancy:0.963 train accurancy:0.96647274
step:12 test accurancy:0.9636 train accurancy:0.9667091
step:13 test accurancy:0.9637 train accurancy:0.9666727
step:14 test accurancy:0.964 train accurancy:0.9666727
step:15 test accurancy:0.9648 train accurancy:0.96714544
step:16 test accurancy:0.9639 train accurancy:0.96672726
step:17 test accurancy:0.9662 train accurancy:0.9674182
step:18 test accurancy:0.9649 train accurancy:0.9673273
step:19 test accurancy:0.9636 train accurancy:0.96736366